File: rotate-matrix.c
#include <stdio.h>
#include <stdlib.h>

/**
 * Small program to rotate a matrix to 90 degrees to the right in place.
 *
 * Ex:
 *
 * Original:
 * A B C D
 * E F G H
 * I J K L
 * M N O P
 *
 * Rotated 90 degrees:
 * M I E A
 * N J F B
 * O K G C
 * P L H D
 *
 */

void print_matrix(char matrix[4][4])
{
    int i = 0, j = 0;

    for(i=0; i < 4; i++)
    {
        for(j=0; j < 4; j++)
            printf("%c\t", matrix[i][j]);
        printf("\n");
    }
    printf("\n");
}

int main()
{
    int i, j, count = 0;
    char tmp;
    char matrix[4][4];

    char letters[16] = {'A', 'B', 'C', 'D', 'E', 'F', 'G', 'H',
                        'I', 'J', 'K', 'L', 'M', 'N', 'O', 'P'};

    // Pre-populate matrix
    for(i=0; i < 4; i++)
    {
        for(j=0;j < 4; j++)
        {
            matrix[i][j] = letters[count];
            count++;
        }
    }

    printf("Original:\n");
    print_matrix(matrix);

    // Reflect matrix down the middle
    for(i=0; i < 4; i++)
    {
        // Switch outer
        tmp = matrix[i][0];
        matrix[i][0] = matrix[i][3];
        matrix[i][3] = tmp;

        // Switch inner
        tmp = matrix[i][1];
        matrix[i][1] = matrix[i][2];
        matrix[i][2] = tmp;
    }
    
    printf("Reflected down the middle y axis:\n");
    print_matrix(matrix);

    // Reflect over right diagonal
    for(i=0; i < 4; i++)
    {
        for(j=3; j >= 0; j--)
        {
            if( i == 0 && j== 2)
            {
                tmp = matrix[1][3];
                matrix[1][3] = matrix[i][j];
                matrix[i][j] = tmp;
            }
            else if( i == 0 && j == 1)
            {
                tmp = matrix[2][3];
                matrix[2][3] = matrix[i][j];
                matrix[i][j] = tmp;
            }
            else if( i == 0 && j == 0)
            {
                tmp = matrix[3][3];
                matrix[3][3] = matrix[i][j];
                matrix[i][j] = tmp;
            }
            else if( i == 1 && j == 1)
            {
                tmp = matrix[2][2];
                matrix[2][2] = matrix[i][j];
                matrix[i][j] = tmp;
            }
            else if( i == 1 && j == 0)
            {
                tmp = matrix[3][2];
                matrix[3][2] = matrix[i][j];
                matrix[i][j] = tmp;
            }
            else if( i == 2 && j == 0)
            {
                tmp = matrix[3][1];
                matrix[3][1] = matrix[i][j];
                matrix[i][j] = tmp;
            }
            else
                continue;
        }
    }
    printf("Reflected over the right diagonal:\n");
    print_matrix(matrix);

    return;
}

File: rotate-matrix.py
def format_matrix(matrix):
    for posX in range(4):
        print "%-2c %-2c %-2c %-2c" % (matrix[posX][0], matrix[posX][1], matrix[posX][2], matrix[posX][3])

def reflect_over_y_axis(matrix):
    for posX in range(4):
        # Switch outer
        tmp = matrix[posX][0]
        matrix[posX][0] = matrix[posX][3]
        matrix[posX][3] = tmp

        # Switch inner
        tmp = matrix[posX][1]
        matrix[posX][1] = matrix[posX][2]
        matrix[posX][2] = tmp

def reflect_over_right_diagonal(matrix):
    for posX in range(4):
        for posY in range(4):
            if posX == 0 and posY == 0:
                tmp = matrix[3][3]
                matrix[3][3] = matrix[posX][posY]
                matrix[posX][posY] = tmp
            elif posX == 1 and posY == 0:
                tmp = matrix[3][2]
                matrix[3][2] = matrix[posX][posY]
                matrix[posX][posY] = tmp
            elif posX == 2 and posY == 0:
                tmp = matrix[3][1]
                matrix[3][1] = matrix[posX][posY]
                matrix[posX][posY] = tmp
            elif posX == 0 and posY == 1:
                tmp = matrix[2][3]
                matrix[2][3] = matrix[posX][posY]
                matrix[posX][posY] = tmp
            elif posX == 1 and posY == 1:
                tmp = matrix[2][2]
                matrix[2][2] = matrix[posX][posY]
                matrix[posX][posY] = tmp
            elif posX == 0 and posY == 2:
                tmp = matrix[1][3]
                matrix[1][3] = matrix[posX][posY]
                matrix[posX][posY] = tmp

if __name__ == "__main__":
    # Initialize 4x4 matrix
    matrix = [[0]*4 for i in range(4)]
    
    letters = ['A', 'B', 'C', 'D', 'E', 'F', 'G', 'H', 'I', 'J', 'K', 'L', 'M', 'N', 'O', 'P']
    currIndex = 0

    for posX in range(4):
        for posY in range(4):
            matrix[posX][posY] = letters[currIndex]
            currIndex += 1

    print "Original:"
    format_matrix(matrix)

    reflect_over_y_axis(matrix)
    print "Reflected over y axis:"
    format_matrix(matrix)
    
    reflect_over_right_diagonal(matrix)
    print "Reflected over right diagonal:"
    format_matrix(matrix)

